File: src/checkers-cpp/StudentAI.cpp
#include "StudentAI.h"
#include <random>

//The following part should be completed by students.
//The students can modify anything except the class name and exisiting functions and varibles.
StudentAI::StudentAI(int col,int row,int p)
	:AI(col, row, p)
{
    board = Board(col,row,p);
    board.initializeGame();
    player = 2;
}

int StudentAI::countCaptures(const Move& move) {
    int captureCount = 0;
    for (int i = 0; i < move.seq.size() - 1; ++i) {
        if (abs(move.seq[i][0] - move.seq[i + 1][0]) > 1 && abs(move.seq[i][1] - move.seq[i + 1][1]) > 1) {
            captureCount++;
        }
    }
    return captureCount;
}

Move StudentAI::GetMove(Move move)
{
    if (move.seq.empty())
    {
        player = 1;
    } else{
        board.makeMove(move,player == 1?2:1);
    }
    vector<vector<Move> > moves = board.getAllPossibleMoves(player);

    Move bestMove;
    int maxCaptures = 0;

    for (const auto& checker_moves : moves) {
        for (const auto& m : checker_moves) {
            if (m.isCapture()) {
                int captureCount = countCaptures(m);
                if (captureCount > maxCaptures) {
                    maxCaptures = captureCount;
                    bestMove = m;
                }
            }
        }
    }

    if (maxCaptures > 0) {
        board.makeMove(bestMove, player);
        return bestMove;
    }

    int i = rand() % (moves.size());
    vector<Move> checker_moves = moves[i];
    int j = rand() % (checker_moves.size());
    Move res = checker_moves[j];
    board.makeMove(res,player);
    return res;
}

File: src/checkers-python/StudentAI.py
from random import randint
from BoardClasses import Move
from BoardClasses import Board
#The following part should be completed by students.
#Students can modify anything except the class name and exisiting functions and varibles.
class StudentAI():

    def __init__(self,col,row,p):
        self.col = col
        self.row = row
        self.p = p
        self.board = Board(col,row,p)
        self.board.initialize_game()
        self.color = ''
        self.opponent = {1:2,2:1}
        self.color = 2
    def get_move(self, move):
        if len(move) != 0:
            self.board.make_move(move, self.opponent[self.color])
        else:
            self.color = 1

        moves = self.board.get_all_possible_moves(self.color)

        # Implement logic to find the move with the longest sequence
        # For now, it's selecting a random move
        if moves:
            longest_move = max(moves, key=lambda group: max(len(m) for m in group))
            chosen_move = max(longest_move, key=len)
            self.board.make_move(chosen_move, self.color)
            return chosen_move

        # Fallback to random move if no other move is found
        index = randint(0, len(moves) - 1)
        inner_index = randint(0, len(moves[index]) - 1)
        move = moves[index][inner_index]
        self.board.make_move(move, self.color)
        return move
        